File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    static boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        int answer = 0;
        int len = dungeons.length;
        
        visited = new boolean[len];
        for(int i=0; i<len; i++){
            if(dungeons[i][0] <= k){
                visited[i] = true;
                answer = Math.max(dfs(len, 1, dungeons, k-dungeons[i][1]),answer);
                visited[i] = false;
            }
        }
        
        return answer;
    }
    
    public int dfs(int n, int cnt, int[][] arr, int f){
        int max = cnt;
        for(int i=0; i<n; i++){
            if(!visited[i] && arr[i][0] <= f){
                visited[i] = true;
                max = Math.max(dfs(n, cnt + 1, arr, f - arr[i][1]), max);
                visited[i] = false;
            }
        }
        
        return max;
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations
import math

def solution(k, dungeons):
    n = len(dungeons)
    answer = 0
    
    line = [i for i in range(n)]
    
    for i in permutations(line, n):
        tmp = k
        count = 0
        
        for j in i:
            if tmp >= dungeons[j][0]:
                tmp -= dungeons[j][1]
                count += 1
        
        answer = max(answer, count)
    
    return answer